### PR TITLE
Add span for connection phase

### LINKED ISSUE #1133: Instrument SQLAlchemy engine connection phase
** Is your feature request related to a problem? **
The SQLAlchemy instrumentation does not trace the actual connection to the database

**Describe the solution you'd like**
I want that `connect` function will also be traced

**Describe alternatives you've considered**
Which alternative solutions or features have you considered?

**Additional context**
We are working with SQLAlchemy (snowflake db) and we implemented a solution where we can see the `connect` span also as attached in the screenshot (the span is `database-connect`)
image

### PR DESCRIPTION
**Pull request by shahargl**
# Description

I've added the code that adds https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1133

Fixes #1133 
https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1133

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

# Does This PR Require a Core Repo Change?

- [ ] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [ ] Followed the style guidelines of this project
- [ ] Changelogs have been updated
- [ ] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

**Review by ocelotl — COMMENTED** [review]
Thanks for the contribution! We need at least one test case that validates these changes are correct :slightly_smiling_face:

**Comment by shahargl** [comment]
@ocelotl no problem! 

> We need at least one test case that validates these changes are correct 
Can you guide me through an example? You mean UT?

**Comment by shahargl** [comment]
@ocelotl 
Added also a test for connection phase

**Comment by srikanthccv** [comment]
you probably wanted to add listen on connect similar to this https://github.com/open-telemetry/opentelemetry-python-contrib/blob/80969a06da77d1e616124de178d12a1ebe3ffe7f/instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py#L87-L91

**Comment by shahargl** [comment]
@srikanthccv 
Hey! Thanks for jumping in :)

So - initially, I thought about this approach too. Still, the problem is that if I add a listener to the `connect` event, we will have only the time the connection is initialized. Still, without having the "connection completed" event, we won't be able to "finish" the span.

Unfortunately, `sqlalchemy` doesn't have this event (https://docs.sqlalchemy.org/en/14/core/events.html)
So I choose to patch the `connect` function so the full connection time will be traced. wdyt?

**Comment by srikanthccv** [comment]
@shahargl makes sense

**Comment by shahargl** [comment]
@ocelotl updated the code to allow SQLAlchemy 1.1 too

File: instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py [hunk 1]
```diff
@@ -77,6 +77,17 @@ def _wrap_create_engine_internal(func, module, args, kwargs):
     return _wrap_create_engine_internal
 
 
+def _wrap_connect(tracer_provider=None):
+    def _wrap_connect_internal(func, module, args, kwargs):
+        tracer = trace.get_tracer(
+            module.name,
+            __version__,
+            tracer_provider=tracer_provider,
+        )
```
Comment (srikanthccv) [inline instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py:86]: Shouldn't be creating a new tracer everytime

File: instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py [hunk 2]
```diff
@@ -77,6 +77,17 @@ def _wrap_create_engine_internal(func, module, args, kwargs):
...
+        with tracer.start_as_current_span("database-connect", kind=trace.SpanKind.CLIENT) as span:
```
Comment (srikanthccv) [inline instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py:87]:
```suggestion
        with tracer.start_as_current_span("connect", kind=trace.SpanKind.CLIENT) as span:
```

File: instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py (hunk 1 above)
Comment (shahargl) [reply instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py:86]: How does it create a new tracer? 

Is it different then the other wrappers (e.g. `_wrap_create_engine`) which use `_get_tracer` (which in turn use `trace.get_tracer` too)? Because that's where I took the inspiration from.

Comment (srikanthccv) [reply instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py:86]: >Is it different then the other wrappers (e.g. _wrap_create_engine)

People don't really create engines all the time. But on the other hand the `connect` is very commonly used and it's unnecessary to create tracer on each `connect`. Create a tracer at the top of module and use it or pass it your wrapper function if it's already created a the wrap call point.

Comment (shahargl) [reply instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py]: Thanks.

I've moved the tracer creation outside of the "connect" wrapped function.

File: instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py [hunk 3]
```diff
@@ -77,6 +77,19 @@ def _wrap_create_engine_internal(func, module, args, kwargs):
...
 
+def _wrap_connect(tracer_provider=None):
+    tracer = trace.get_tracer(
+        _instrumenting_module_name,
+        __version__,
+        tracer_provider=tracer_provider,
+    )
+
+    def _wrap_connect_internal(func, module, args, kwargs):
+        with tracer.start_as_current_span("database-connect", kind=trace.SpanKind.CLIENT) as span:
```
Comment (srikanthccv) [inline instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py:88]:
```suggestion
        with tracer.start_as_current_span("connect", kind=trace.SpanKind.CLIENT) as span:
```

File: instrumentation/opentelemetry-instrumentation-sqlalchemy/tests/test_sqlalchemy.py [hunk 4]
```diff
@@ -85,11 +101,15 @@ async def run():
             async with engine.connect() as cnx:
                 await cnx.execute(sqlalchemy.text("SELECT	1 + 1;"))
             spans = self.memory_exporter.get_finished_spans()
-            self.assertEqual(len(spans), 1)
-            self.assertEqual(spans[0].name, "SELECT :memory:")
+            self.assertEqual(len(spans), 2)
+            # first span - the connection to the db
+            self.assertEqual(spans[0].name, "database-connect")
```
Comment (srikanthccv) [inline instrumentation/opentelemetry-instrumentation-sqlalchemy/tests/test_sqlalchemy.py:106]: You need to update these refs as well

**Review by ocelotl — APPROVED** [review]

File: instrumentation/opentelemetry-instrumentation-sqlalchemy/tests/test_sqlalchemy.py (hunk 4 above)
Comment (shahargl) [reply instrumentation/opentelemetry-instrumentation-sqlalchemy/tests/test_sqlalchemy.py]: Got you, fixed

**Comment by shahargl** [comment]
@ocelotl @srikanthccv - fixed the tests :) feel free to approve/merge

**Comment by ocelotl** [comment]
> @ocelotl @srikanthccv - fixed the tests :) feel free to approve/merge

I see some tests failing, please take a look, @shahargl

**Comment by shahargl** [comment]
@ocelotl fixed, can you report the status so the tests will run?

File: dev-requirements.txt [hunk 5]
```diff
@@ -7,7 +7,7 @@ mypy==0.790
 sphinx
 sphinx-rtd-theme~=0.4
 sphinx-autodoc-typehints
-pytest!=5.2.3
+# pytest!=5.2.3
```
Comment (srikanthccv) [inline dev-requirements.txt:10]: Why is this changed?

Comment (srikanthccv) [reply dev-requirements.txt:10]: Ah I get it, somehow I slept through all of the drama happened over atomicwrites package.

Comment (shahargl) [reply dev-requirements.txt:10]: do you mean this? https://github.com/untitaker/python-atomicwrites/issues/61

Comment (srikanthccv) [reply dev-requirements.txt:10]: Please revert this since the packages are uploaded.

File: CHANGELOG.md [hunk 6]
```diff
@@ -52,6 +52,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   ([#1110](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1110))
 - Integrated sqlcommenter plugin into opentelemetry-instrumentation-django
   ([#896](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/896))
+- `opentelemetry-instrumentation-sqlalchemy` Added span for the connection phase ([#1133](https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1133))
```
Comment (srikanthccv) [inline CHANGELOG.md:55]: Move this to unreleased section

File: instrumentation/opentelemetry-instrumentation-sqlalchemy/tests/test_sqlalchemy.py [hunk 7]
```diff
@@ -46,9 +46,13 @@ def test_trace_integration(self):
         cnx.execute("SELECT	1 + 1;").fetchall()
         spans = self.memory_exporter.get_finished_spans()
 
-        self.assertEqual(len(spans), 1)
-        self.assertEqual(spans[0].name, "SELECT :memory:")
+        self.assertEqual(len(spans), 2)
+        # first span - the connection to the db
+        self.assertEqual(spans[0].name, "connect")
         self.assertEqual(spans[0].kind, trace.SpanKind.CLIENT)
+        # second span - the query it self
```
Comment (srikanthccv) [inline instrumentation/opentelemetry-instrumentation-sqlalchemy/tests/test_sqlalchemy.py:53]:
```suggestion
        # second span - the query itself
```

**Review by srikanthccv — APPROVED** [review]
LGTM. Minor comments.

File: instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py [hunk 8]
```diff
@@ -77,6 +77,22 @@ def _wrap_create_engine_internal(func, module, args, kwargs):
...
 
 
+def _wrap_connect(tracer_provider=None):
+    tracer = trace.get_tracer(
+        _instrumenting_module_name,
+        __version__,
+        tracer_provider=tracer_provider,
+    )
+
+    def _wrap_connect_internal(func, module, args, kwargs):
```
Comment (ocelotl) [inline instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py:87]: This is causing a lint failure, please add a disabling comment.

**Comment by shahargl** [comment]
@ocelotl 
Thanks again, fixed your comments. Sorry it takes too long, can you rerun the tests?

**Comment by shahargl** [comment]
@ocelotl 
4 tests failing with 
```
17 | #include "codecs.h"
          |          ^~~~~~~~~~
    compilation terminated.
    error: command 'gcc' failed with exit status 1
```
So idk why, do you have idea?

Also, fixed the `docker-tests`

File: tox.ini [hunk 9]
```diff
@@ -205,7 +205,7 @@ envlist =
 
 [testenv]
 deps =
-  -c dev-requirements.txt
+  -r dev-requirements.txt
```
Comment (srikanthccv) [inline tox.ini:208]: Why was this needed?

Comment (shahargl) [reply tox.ini]: my bad, it was something while debugging locally, reverted

Comment (srikanthccv) [reply tox.ini]: This was the reason for pypy build failures
